presto-main/src/main/java/com/facebook/presto/cost/AggregationStatsRule.java

```diff
@@ -76,27 +94,59 @@ public static PlanNodeStatsEstimate groupBy(PlanNodeStatsEstimate sourceStats, C
             }));
         }
 
-        double rowsCount = 1;
+        double rowsCount = getDefaultAggregateSelectivityCoefficient(session);
+        double totalSize = 0;
         for (VariableReferenceExpression groupByVariable : groupByVariables) {
             VariableStatsEstimate symbolStatistics = sourceStats.getVariableStatistics(groupByVariable);
             int nullRow = (symbolStatistics.getNullsFraction() == 0.0) ? 0 : 1;
             rowsCount *= symbolStatistics.getDistinctValuesCount() + nullRow;
+            if (!isNaN(symbolStatistics.getAverageRowSize())) {
+                totalSize += symbolStatistics.getAverageRowSize() * rowsCount;
+            }
+        }
+        // Set estimation boundary for aggregation:
+        // If the source only return a single row, it's guaranteed the upper bound is a single row output for this agg.
+        // Aggregation node cannot return output row count more than estimated row count of source node.
+        // So if source stats is present, use the output row count from the source stats as the estimated output row
+        // count as the upper bound limit to avoid returning an unknown output potentially breaks the join search space
+        double outputRowCountFromSource = getDefaultAggregateSelectivityCoefficient(session) * sourceStats.getOutputRowCount();
+        if (!isNaN(rowsCount)) {
+            result.setOutputRowCount(min(rowsCount, outputRowCountFromSource));
+        }
+        else {
+            // groupBy key may be a derived expression which may not have associated stats hence return NaN.
+            // In such case, use source stats to bound the output row count
+            result.setOutputRowCount(outputRowCountFromSource);
         }
-        result.setOutputRowCount(min(rowsCount, sourceStats.getOutputRowCount()));
 
         for (Map.Entry<VariableReferenceExpression, Aggregation> aggregationEntry : aggregations.entrySet()) {
-            result.addVariableStatistics(aggregationEntry.getKey(), estimateAggregationStats(aggregationEntry.getValue(), sourceStats));
+            VariableStatsEstimate aggVariableStatsEstimate = estimateAggregationStats(aggregationEntry.getValue(), sourceStats, rowsCount);
+            result.addVariableStatistics(aggregationEntry.getKey(), aggVariableStatsEstimate);
+            totalSize += aggVariableStatsEstimate.getAverageRowSize() * aggVariableStatsEstimate.getDistinctValuesCount();
```

Comment:
I'm confused about the totalsize calculation in general here.  I think the totalsize should something like the sum of the average size per-column (including both aggregation and group by columns) * the number of rows

Reply:
you are right. so the implementation will be changed to add up all the average per-column size from group-by variables first, and then sum it with the avg size from aggregation columns. Overall sum is multiplied by the number of rows. I add a comment to elaborate the logic.